File: packages/shared/appearance.ts
import type { SidebarThemeSetting } from './session-grid-contract-core';

export type ColorScheme = 'light' | 'dark';
export type ContentThemeSetting = ColorScheme | 'system' | 'app';

export function normalizeContentThemeSetting(value: unknown): ContentThemeSetting {
  return value === 'light' || value === 'dark' || value === 'system' ? value : 'app';
}

/**
 * CDXC:Theming 2026-09-13 DECISION:
 * User: chat and terminal follow the app theme by default, with optional overrides in one Theme section at the top of Settings.
 * This supersedes separate System chat and Dark terminal defaults; explicit saved overrides and the existing dark palettes remain unchanged.
 * SEE-ALSO: apps/desktop/src/shared_settings.rs resolves the same settings for native hosts.
 */
export function resolveContentThemeSetting(
  setting: ContentThemeSetting,
  appTheme: SidebarThemeSetting
): ColorScheme | 'system' {
  if (setting !== 'app') return setting;
  if (appTheme === 'system') return 'system';
  return appTheme === 'plain-light' ? 'light' : 'dark';
}

File: apps/desktop/src/shared_settings/appearance.rs
use serde_json::{Map, Value};

/// CDXC:Theming 2026-09-13 SEE-ALSO:
/// packages/shared/appearance.ts owns the Follow app default and explicit Light, Dark, and System overrides shared with React.
pub fn effective_content_color_scheme(
    settings: &Map<String, Value>,
    setting_key: &str,
) -> &'static str {
    match settings.get(setting_key).and_then(Value::as_str) {
        Some("light") => "light",
        Some("dark") => "dark",
        Some("system") => "system",
        _ => match settings.get("sidebarTheme").and_then(Value::as_str) {
            Some("plain-light") => "light",
            Some("system") | None => "system",
            _ => "dark",
        },
    }
}
